DESIGN.md

```diff
@@ -0,0 +1,126 @@
+1. Resource control - since the user is able to limit the task resources we need to have a small layer for working with the file system.  
+
+### Resource control
+Three subsystems will be used for allowing a user to limit a job:
+1.  blkio - parameter _weight_ is used.
+1.  cpu - parameter _shares_ is used. 
+1. memory - parameter _limit_in_bytes_ is used. It sets the upper limit of memory available to a particular job. 
+
+Initially, we set up a **teleworker** group with blkio.weight and cpu.shares parameters set to 1000 (i.e. maximum).
+Every job launched without limiting any of the parameters will be placed directly to **cgroup.procs** of that group, so basically we will write the job PID into the following:
+```
+/sys/fs/cgroup/cpu/teleworker/cgroup.procs
+/sys/fs/cgroup/blkio/teleworker/cgroup.procs
+/sys/fs/cgroup/memory/teleworker/cgroup.procs
+```
+If a user provides any of the limits when creating a task, we create a new group inside of the _teleworker_ named by the ID of that task. For consistency we do that for all the resources (i.e. the group _cpu/teleworker/uuid_ will be created even if only memory and io limits were set by the user). This means that we write PID of any limited job to
+```
+/sys/fs/cgroup/<resource>/teleworker/<uuid>/cgroup.procs
+```
+
+When the job is finished or terminated we also remove the related directories.
+
+
+## Server
+
+Server will extensively use the library and will probably have only some basic auth logic (see [Security](#security)) and a layer of logic for storing the current system state in memory.
+
+Basic flow of starting a new job looks like this:
+
+1. Server receives the _start_ command sent by a user with at least _command_ provided to execute.
+1. User is authorized and we extract the login data.
+1. With _exec.Command_ we create the command structure using optional arguments if provided in the request.
+1. We need to process the output of the task, so the following is required (both for stdout and stderr):
+
+    * A pipe is created, it will be connected to the command's out when the command starts.
+    * We set up a broker and a buffer that will contain all the data produced by the task.
+    * When the command will start we set up two goroutines - in one of them we read the pipe and publish it's content to the broker. Another one receives the messages from broker and writes them into the buffer.
+
+1. The user login is set as a creator of the task. While later managing the task, it will be decided if the user is legible to perform requests based on the login assigned to the task.
+1. **cgexec** tool will be used in order to run the task in the related cgroup - _/teleworker_ if no limits were provided, or _teleworker/UUID_ if at least one of mem/cpu/io limits is set by the user.
```

Comment:
Can you implement this without `cgexec`?

Reply:
I've come up with the idea to run `exec.Command` with `/proc/self/exe` as the main command and providing all the other arguments (including the user task itself). More details are provided in the doc.
It should solve the problem (i.e. the pid will be added to the group before the user command is executed), but I'm not actually sure that there is no more proper way to solve this. Yet, it seems like trying to solve this with syscalls will result in a lot more work to do.
What do you think?

Reply:
yes this should work, nice.